File: backup/golang/easy/538.把二叉搜索树转换为累加树.go
package easy

/*
 * @lc app=leetcode.cn id=538 lang=golang
 *
 * [538] 把二叉搜索树转换为累加树
 *
 * https://leetcode-cn.com/problems/convert-bst-to-greater-tree/description/
 *
 * algorithms
 * Easy (56.10%)
 * Likes:    147
 * Dislikes: 0
 * Total Accepted:    8.3K
 * Total Submissions: 14.7K
 * Testcase Example:  '[5,2,13]'
 *
 * 给定一个二叉搜索树（Binary Search Tree），把它转换成为累加树（Greater
 * Tree)，使得每个节点的值是原来的节点值加上所有大于它的节点值之和。
 *
 * 例如：
 *
 *
 * 输入: 二叉搜索树:
 * ⁠             5
 * ⁠           /   \
 * ⁠          2     13
 *
 * 输出: 转换为累加树:
 * ⁠            18
 * ⁠           /   \
 * ⁠         20     13
 *
 *
 */
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func convertBST(root *TreeNode) *TreeNode {
	// 从数的右下角开始累加
	sum := 0
	visit := func(node *TreeNode) {
		node.Val += sum
		sum = node.Val
	}

	var lastOrder func(node *TreeNode)
	lastOrder = func(node *TreeNode) {
		if node == nil {
			return
		}
		lastOrder(node.Right)
		visit(node)
		lastOrder(node.Left)
	}
	lastOrder(root)
	return root
}

File: labuladong/二叉搜索树/538.把二叉搜索树转换为累加树.py

#
# @lc app=leetcode.cn id=538 lang=python3
#
# [538] 把二叉搜索树转换为累加树
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def convertBST(self, root):
        """将bst从大到小遍历，既后序遍历"""
        self.sum = 0
        self.traverse(root)
        return root

    def traverse(self, root):
        if not root:
            return
        self.traverse(root.right)
        self.sum += root.val
        root.val = self.sum
        self.traverse(root.left)


# @lc code=end

